### PR TITLE
Added tuple2 - Map.Entry conversion methods

### PR DESCRIPTION
**Pull request by ashrko619**
PR for #1545

### CODE REVIEW

File: javaslang/generator/Generator.scala [hunk 1]
```diff
@@ -1330,6 +1346,18 @@ def generateMainClasses(): Unit = {
...
               return new Tuple$i<>($params);
           }
+          ${(i == 2).gen(xs"""
+
+           /**
+           * Creates a tuple of ${i.numerus("element")} from java.util.Map.Entry.
+           * @param entry A java.util.Map.Entry
+           * @return a tuple of ${i.numerus("element")}.
+           */
+            static <$generics> Tuple$i<$generics> fromEntry(Map.Entry<T1, T2> entry) {
```
Comment (ruslansennov) [inline javaslang/generator/Generator.scala]: Please use `$generics` instead of `T1, T2`

File: javaslang/generator/Generator.scala [hunk 2]
```diff
@@ -1119,6 +1123,17 @@ def generateMainClasses(): Unit = {
...
               }
+
+             /$javadoc
+             * Converts the tuple to java.util.Map.Entry {@code Tuple}.
+             *
+             * @return A  java.util.Map.Entry where the first element is the key and the second
+             * element is the value.
+             */
+              public Map.Entry<T1,T2> toEntry(){
+                  return new AbstractMap.SimpleEntry<>(_1, _2);
```
Comment (ruslansennov) [inline javaslang/generator/Generator.scala]: just question: why you use `SimpleEntry` but not `SimpleImmutableEntry`?

File: javaslang/generator/Generator.scala [hunk 3]
```diff
@@ -1957,6 +1990,14 @@ def generateTestClasses(): Unit = {
                 public void shouldSwap() {
                     $assertThat(createIntTuple(1, 2).swap()).isEqualTo(createIntTuple(2, 1));
                 }
+
+                @$test
+                public void shouldConvertToEntry() {
+                    Tuple2<Integer, Integer> tuple= createIntTuple(1,2);
```
Comment (ruslansennov) [inline javaslang/generator/Generator.scala]: `Tuple$i$generics`

File: javaslang/generator/Generator.scala [hunk 4]
```diff
@@ -1957,6 +1990,14 @@ def generateTestClasses(): Unit = {
...
+                    Map.Entry<Integer,Integer> entry = new AbstractMap.SimpleEntry<>(1, 2);
```
Comment (ruslansennov) [inline javaslang/generator/Generator.scala]: Map.Entry$generics

File: javaslang/generator/Generator.scala (hunk 2 above)
Comment (ashrko619) [inline javaslang/generator/Generator.scala]: I wasn't sure on what to use, so went with ```SimpleEntry``` - will change it to ```SimpleImmutableEntry``` :)

Comment (danieldietrich) [inline javaslang/generator/Generator.scala]: @ashrko619 @ruslansennov Let's return a `SimpleEntry`.

Javaslang is immutable-by-design. The `java.util` package is a mutable-by-default parallel universe. The `Map.Entry` has a `setValue()` method. Returning an object that throws would be unsafe.

Thx for asking and giving an alternative. 👏

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 5]
```diff
@@ -95,6 +96,15 @@ static Tuple0 empty() {
     }
 
     /**
+    * Creates a tuple of two elements from java.util.Map.Entry.
+    * @param entry A java.util.Map.Entry
+    * @return a tuple of two elements.
+    */
+     static <T1, T2> Tuple2<T1, T2> fromEntry(Map.Entry<T1, T2> entry) {
```
Comment (danieldietrich) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: I think we can widen the argument types:

```java
static <T1, T2> Tuple2<T1, T2> fromEntry(Map.Entry<? extends T1, ? extends T2> entry) { ...
```

Only immutable types are covariant in generic arguments, here is the classical counter example (taken from [StackOverflow](http://stackoverflow.com/a/18666878/1110815)):

```java
// Illegal code - because otherwise life would be Bad
List<Dog> dogs = new List<Dog>();
List<Animal> animals = dogs; // Awooga awooga
animals.add(new Cat());
Dog dog = dogs.get(0); // This should be safe, right?
```

Projected to our use-case, this can't happen:

```java
// Illegal code - because otherwise life would be Bad
Map.Entry<T, Dog> t_dog = new SimpleEntry<>(t, dog);
Tuple<T, Animal> t_animal = Tuple.fromEntry(entry);
Tuple<T, Animal> t_animal2 = t_animal.update2(new Cat());
Dog dog = t_dog.getValue(); // ok
```

**Side-note:** We can improve the generics of other types, too. We will do it in another PR.

```java
public interface Validation<E, T> extends Value<T> {
    // should be `Either<? extends E, ? extends T> either`
    static <E, T> Validation<E, T> fromEither(Either<E, T> either) { ... }
}
```

Comment (ashrko619) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: _Only immutable types are covariant in generic arguments_ - Nice 👍 
@danieldietrich will make the changes and thanks for the example 😄

File: javaslang/src-gen/test/java/javaslang/Tuple2Test.java [hunk 6]
```diff
@@ -91,6 +93,13 @@ public void shouldSwap() {
     }
 
     @Test
+    public void shouldConvertToEntry() {
+        Tuple2<Integer, Integer> tuple= createIntTuple(1,2);
```
Comment (danieldietrich) [inline javaslang/src-gen/test/java/javaslang/Tuple2Test.java]: minor (space): `tuple =` instead of `tuple=`

File: javaslang/src-gen/main/java/javaslang/Tuple2.java [hunk 7]
```diff
@@ -133,15 +135,25 @@ public T2 _2() {
...
-      return Tuple.of(_2, _1);
-    }
+    * Converts the tuple to java.util.Map.Entry {@code Tuple}.
+    *
+    * @return A  java.util.Map.Entry where the first element is the key and the second
+    * element is the value.
+    */
+     public Map.Entry<T1, T2> toEntry(){
+         return new AbstractMap.SimpleEntry<>(_1, _2);
+     }
```
Comment (danieldietrich) [inline javaslang/src-gen/main/java/javaslang/Tuple2.java]: The changes look good to me, thank you!
We are nearly there.
Before I merge, please align the lines right. The indentation needs to be fixed in the generator.
See also https://github.com/ashrko619/javaslang/blob/26257170f75eac814354e5f27bca6c48a7455a52/javaslang/src-gen/main/java/javaslang/Tuple2.java#L138-L156

**Comment by danieldietrich** [comment]
Great, thank you!
